src/scaling_of_discrete_log.rs

```diff
@@ -0,0 +1,856 @@
+        PLAINTEXT_SPACE_SCALAR_LIMBS,
+        EncryptionKey,
+    >,
+    homomorphic_encryption::RandomnessSpacePublicParameters<
+        PLAINTEXT_SPACE_SCALAR_LIMBS,
+        EncryptionKey,
+    >,
+    homomorphic_encryption::CiphertextSpacePublicParameters<
+        PLAINTEXT_SPACE_SCALAR_LIMBS,
+        EncryptionKey,
+    >,
+    homomorphic_encryption::PublicParameters<PLAINTEXT_SPACE_SCALAR_LIMBS, EncryptionKey>,
+    homomorphic_encryption::CiphertextSpaceValue<PLAINTEXT_SPACE_SCALAR_LIMBS, EncryptionKey>,
+>;
+impl<
+        const PLAINTEXT_SPACE_SCALAR_LIMBS: usize,
+        const SCALAR_LIMBS: usize,
+        GroupElement: PrimeGroupElement<SCALAR_LIMBS>,
+        EncryptionKey: AdditivelyHomomorphicEncryptionKey<PLAINTEXT_SPACE_SCALAR_LIMBS>,
+    > maurer::Language<SOUND_PROOFS_REPETITIONS>
+    for Language<PLAINTEXT_SPACE_SCALAR_LIMBS, SCALAR_LIMBS, GroupElement, EncryptionKey>
+where
+    Uint<PLAINTEXT_SPACE_SCALAR_LIMBS>: Encoding,
+{
+    type WitnessSpaceGroupElement =
+        WitnessSpaceGroupElement<PLAINTEXT_SPACE_SCALAR_LIMBS, EncryptionKey>;
+
+    type StatementSpaceGroupElement = crate::encryption_of_discrete_log::StatementSpaceGroupElement<
+        PLAINTEXT_SPACE_SCALAR_LIMBS,
+        SCALAR_LIMBS,
+        GroupElement,
+        EncryptionKey,
+    >;
+
+    type PublicParameters =
+        PublicParameters<PLAINTEXT_SPACE_SCALAR_LIMBS, SCALAR_LIMBS, GroupElement, EncryptionKey>;
+
+    const NAME: &'static str = "Scaling of a Discrete Log";
+
+    fn homomorphose(
+        witness: &Self::WitnessSpaceGroupElement,
+        language_public_parameters: &Self::PublicParameters,
+    ) -> maurer::Result<Self::StatementSpaceGroupElement> {
+        if SCALAR_LIMBS >= PLAINTEXT_SPACE_SCALAR_LIMBS {
+            return Err(Error::InvalidPublicParameters);
+        }
+
+        let group_order = Uint::<PLAINTEXT_SPACE_SCALAR_LIMBS>::from(
+            &GroupElement::Scalar::order_from_public_parameters(
+                &language_public_parameters.scalar_group_public_parameters,
+            ),
+        );
+
+        let generator = GroupElement::generator_from_public_parameters(
+            language_public_parameters.group_public_parameters(),
+        )?;
+
+        let base_by_discrete_log = generator.scalar_mul(&witness.discrete_log().value().into());
+
+        let encryption_key =
+            EncryptionKey::new(&language_public_parameters.encryption_scheme_public_parameters)
+                .map_err(|_| maurer::Error::InvalidPublicParameters)?;
+
+        let ciphertext = homomorphic_encryption::CiphertextSpaceGroupElement::<
+            PLAINTEXT_SPACE_SCALAR_LIMBS,
+            EncryptionKey,
+        >::new(
+            language_public_parameters.ciphertext,
+            language_public_parameters
+                .encryption_scheme_public_parameters
+                .ciphertext_space_public_parameters(),
+        )?;
+
+        // no mask needed, as we don't need circuit privacy.
```

Comment:
Now that I'm reading it, I think we should be a bit more careful/clear:
```suggestion
        // No masking of the plaintext is needed, as we don't need secure function evaluation.
        // However, we do want to re-randomize the ciphertext when doing the scalar multiplication, to ensure circuit privacy against an adversary that does not hold the private key, that is, the centralised party A.
```
I think the same comment should also be changed in EncDH if I remember correctly.